programs/ziti-edge-tunnel/instance-config.c

```diff
@@ -0,0 +1,151 @@
+            }
+        }
+        fclose(config_file);
+    } else {
+        ZITI_LOG(INFO, "The config file %s does not exist. This is normal if this is a new install or if the config file was removed manually", config_file_name);
+    }
+    return loaded;
+}
+
+bool load_tunnel_status_from_file(uv_loop_t* ziti_loop) {
+    char* config_path = get_system_config_path();
+
+    uv_fs_t fs;
+    int check = uv_fs_mkdir(ziti_loop, &fs, config_path, 0755, NULL);
+    if (check == 0) {
+        ZITI_LOG(TRACE,"config path is created at %s", config_path);
+    } else {
+        ZITI_LOG(TRACE,"config path is found at %s", config_path);
+    }
+    bool loaded = false;
+
+    char* config_file_name = get_config_file_name(config_path);
+    char* bkp_config_file_name = get_backup_config_file_name(config_path);
+
+    // try to load tunnel status from config file
+    loaded = load_config_from_file(config_file_name);
+
+    // try to load tunnel status from backup config file
+    if (!loaded) {
+        loaded = load_config_from_file(bkp_config_file_name);
+    }
+
+    // not able to load the tunnel status from both the config and backup files
+    if (!loaded) {
+        ZITI_LOG(WARN, "Config files %s and the backup file cannot be read or they do not exist, will create a new config file or the old one will be overwritten", config_file_name);
+    }
+
+    free(config_file_name);
+    free(bkp_config_file_name);
+    free(config_path);
+    return loaded;
+}
+
+bool save_tunnel_status_to_file() {
+    tunnel_status* stat = get_tunnel_status();
+    size_t json_len;
+    char* tunnel_status = tunnel_status_to_json(stat, 0, &json_len);
+    bool saved = false;
+
+    if (json_len > 0) {
+        char* config_path = get_system_config_path();
+
+        char* config_file_name = get_config_file_name(config_path);
+        char* bkp_config_file_name = get_backup_config_file_name(config_path);
+
+        uv_mutex_lock(&mutex);
+        //copy config to backup file
+        int rem = remove(bkp_config_file_name);
+        if (rem == 0) {
+            ZITI_LOG(DEBUG, "Deleted backup config file %s", bkp_config_file_name);
+        }
+        if (rename(config_file_name, bkp_config_file_name) == 0) {
+            ZITI_LOG(DEBUG, "Copied config file to backup config file %s", bkp_config_file_name);
+        } else {
+            ZITI_LOG(DEBUG, "Could not copy config file [%s] to backup config file, the config might not exists at the moment", config_file_name);
+        }
+
+        // write tunnel status to the config file
+        FILE* config = fopen(config_file_name, "w");
+        if (config == NULL) {
+            ZITI_LOG(ERROR, "Could not open config file %s to store the tunnel status data", config_file_name);
+        } else {
+            char* tunnel_status_data = tunnel_status;
+            for (int i =0; i< json_len; i=i+MIN_BUFFER_LEN-1, tunnel_status_data=tunnel_status_data+MIN_BUFFER_LEN-1) {
```

Comment:
I may be missing something but ( think this entire loop could be replaced with:

    fputs(tunnel_status, config);

Reply:
this file will contain multiline data, and as of now it will have more than 1000 lines of data. So I thought, we have to write the data using buffer. If that is not required, I will change it as fputs